lib/autoproj/python.rb

```diff
@@ -0,0 +1,317 @@
+        # @return [String,String] python path and python version
+        def self.activate_python(ws: Autoproj.workspace,
+                                 bin: nil,
+                                 version: nil)
+            bin, version = resolve_python(ws: ws, bin: bin, version: version)
+            ws.config.set('python_executable', bin, true)
+            ws.config.set('python_version', version, true)
+
+            ws.osdep_suffixes << "python#{$1}" if version =~ /^([0-9]+)\./
+
+            rewrite_python_shims(bin, ws.root_dir)
+            rewrite_pip_shims(bin, ws.root_dir)
+            [bin, version]
+        end
+
+        def self.deactivate_python(ws: Autoproj.workspace)
+            remove_python_shims(ws.root_dir)
+            remove_pip_shims(ws.root_dir)
+            ws.config.reset('python_executable')
+            ws.config.reset('python_version')
+        end
+
+        # Allow to update the PYTHONPATH for package if autoproj configuration
+        # USE_PYTHON is set to true.
+        # Then tries to guess the python binary from Autobuild.programs['python']
+        # and system's default setting
+        # @param [Autobuild::Package] pkg
+        # @param [Autoproj.workspace] ws Autoproj workspace
+        # @param [String] bin Path to a custom python version
+        # @param [String] version version constraint for python executable
+        # @return tuple of [executable, version, site-packages path] if set,
+        #    otherwise nil
+        def self.activate_python_path(pkg,
+                                 ws: Autoproj.workspace,
+                                 bin: nil,
+                                 version: nil)
+            return unless ws.config.get('USE_PYTHON', nil)
+
+            bin, version = resolve_python(ws: ws, bin: bin, version: version)
+            path = File.join(pkg.prefix, "lib",
+                             "python#{version}", "site-packages")
+            pkg.env_add_path 'PYTHONPATH', path
+
+            [bin, version, path]
+        end
+
+        def self.prepare_venv(ws: Autoproj.workspace, bin: nil, version: nil,
+                              venv_name: ".autoproj_python_venv")
+
+          python_bin, = resolve_python(ws: ws, bin: bin, version: version)
+          cmd = "#{python_bin} -m pip install --user -U virtualenv"
+          msg, status = Open3.capture2e(cmd)
+          unless status.success?
+              raise "Autoproj::Python.prepare_venv installation of "\
+                  " virtualenv failed: '#{msg}'"
+          end
+
+          cmd = "#{python_bin} -m virtualenv #{venv_name}"
+          msg, status = Open3.capture2e(cmd)
+          unless status.success?
+              raise "Autoproj::Python.prepare_venv preparation of"\
+                  " virtual env '#{venv_name}' failed: '#{msg}'"
+          end
+
+          ws.env.set('VIRTUAL_ENV_DISABLE_PROMPT', 1)
+          ws.env.source_after File.join(Autoproj.root_dir, venv_name, "bin", "activate")
```

Comment:
This, unfortunately, won't integrate well. `Environment#source_after` will just add a source line to the end of the generated `env.sh` file but that file should only be treated as a convenience for the user and not as a dependency. In other words, running `.autoproj/bin/autoproj` should work as expected even if you don't pre-load `env.sh`. This is important for external tools that rely on an autoproj workspace.

If we want to integrate a "virtual env" tool (which I think is a good idea) I think you should look into [pipenv](https://pypi.org/project/pipenv/) which works more like Ruby's Bundler but then it would probably be best to implement a new `PackageManagers::PipenvManager` and deprecate/remove the current `PackageManagers::PipManager`.

Also, since we are already installing pip packages locally in the workspace, what do we get with the virtualenv integration that you implemented?

Reply:
> Also, since we are already installing pip packages locally in the workspace, what do we get with the virtualenv integration that you implemented?

I assumed this is a more 'python'-conforming setup and thus improving the integration, but I do not consider me as the python expert here. I will have a look at pipenv.

Reply:
Yeah, but the current code does not integrate well... So, if there's no reason to use virtualenv I think we should either use something that integrates better or just don't virtualenv.